### PR TITLE
metrics: more logs and metrics for resolved-ts

### PR DESCRIPTION
**Pull request by ekexium**
### What is changed and how it works?

Issue Number: ref #15082

What's Changed:

```commit-message
Add more logs and metrics for resolved-ts.
```

Example logs
```
[2023/08/29 16:48:18.118 +08:00] [INFO] [endpoint.rs:505] ["the max gap of leader resolved-ts is large"] [last_resolve_attempt="Some(LastAttempt { success: false, ts: TimeStamp(44388
8082736381953), reason: \"lock\", lock: Some(7480000000000000625F728000000002512B5C) })"] [duration_to_last_update_safe_ts=10648ms] [min_memory_lock=None] [txn_num=0] [lock_num=0] [m
in_lock=None] [safe_ts=443888117326544897] [gap=110705ms] [region_id=291]
[2023/08/29 16:48:18.118 +08:00] [INFO] [endpoint.rs:526] ["the max gap of follower safe-ts is large"] [oldest_candidate=None] [latest_candidate=None] [applied_index=3276] [duration_
to_last_consume_leader=11460ms] [resolved_ts=443888117117353985] [safe_ts=443888117117353985] [gap=111503ms] [region_id=273]
[2023/08/29 16:48:18.118 +08:00] [INFO] [endpoint.rs:547] ["the max gap of follower resolved-ts is large; it's the same region that has the min safe-ts"]
```
image

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

Side effects

### Release note 

```release-note
More diagnostic logs and metrics are added for resolved-ts.
```

### CODE REVIEW

File: components/resolved_ts/src/resolver.rs [hunk 1]
```diff
@@ -33,9 +66,19 @@ pub struct Resolver {
...
+    pub(crate) last_attempt: Option<LastAttempt>,
+}
+
+#[derive(Clone, Debug)]
+#[allow(unused)]
+pub(crate) struct LastAttempt {
+    success: bool,
+    ts: TimeStamp,
+    reason: String,
+    lock: Option<Key>,
```
[the remark this answers is not in the document]
Comment (ekexium) [reply components/resolved_ts/src/resolver.rs:81]: The `Display` and `Debug` traits of `Key` already respect the log redaction setting, I suppose

File: components/resolved_ts/src/resolver.rs [hunk 2]
```diff
@@ -8,21 +8,54 @@ use tikv_util::{
...
 
 const MAX_NUMBER_OF_LOCKS_IN_LOG: usize = 10;
 const ON_DROP_WARN_HEAP_SIZE: usize = 64 * 1024 * 1024; // 64MB
 
+#[derive(Clone)]
+pub enum TsSource {
+    // A lock in LOCK CF
+    Lock(Key),
+    // A memory lock in concurrency manager
+    MemoryLock(Key),
```
[the remark this answers is not in the document]
Comment (ekexium) [reply components/resolved_ts/src/resolver.rs]: It seems we can hardly avoid cloning keys from concurrency manager. I've used Arc<[u8]> for keys in LOCK CF.

File: components/resolved_ts/src/resolver.rs [hunk 3]
```diff
@@ -259,17 +329,37 @@ impl Resolver {
         }
 
         // Find the min start ts.
-        let min_lock = self.lock_ts_heap.keys().next().cloned();
+        let (min_lock, min_lock_key) = match self.lock_ts_heap.iter().next() {
```
[the remark this answers is not in the document]
Comment (ekexium) [reply components/resolved_ts/src/resolver.rs]: Done. Please take another look

**Review by you06 — APPROVED** [review]

**Comment by cfzjywxk** [comment]
@overvenus PTAL

File: components/resolved_ts/src/resolver.rs [hunk 4]
```diff
@@ -240,10 +302,18 @@ impl Resolver {
     }
 
     /// Try to advance resolved ts.
+    /// `memory_lock`: If the source is from a memory lock in concurrency
+    /// manger, pass the key for diagnosis
+    ///
```
[the remark this answers is not in the document]
Comment (ekexium) [reply components/resolved_ts/src/resolver.rs:307]: Oh yes.

**Review by zyguan — APPROVED** [review]

**Pull request merged**